shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,317 @@
+     * It should normally be coupled with a Causing DSL word.
+     * When coupled with a Causing DSL word and a provider that has a When condition, this entry will run in place of
+     * the provider until the When condition is fulfilled.
+     * If it is used without a Causing DSL word it will run once as a provider before the actual provider is executed.
+     *
+     * @tparam T the provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Entering : public commands::ProviderBase<T, commands::ProviderAction::ENTERING> {};
+
+    /**
+     * This DSL word is used to define a leaving transition from a provider.
+     * It should normally be combined with a Causing DSL word.
+     * In this case when another provider needs control and uses a When condition, this will run until that
+     * condition is fulfilled. A Leaving condition will be preferred over an Entering condition that provides the
+     * same Causing statement (the system will rather a module cleans up after itself than ask another module to do
+     * so).
+     *
+     * @tparam T the provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Leaving : public commands::ProviderBase<T, commands::ProviderAction::LEAVING> {};
+
+    /**
+     * This DSL word is used to define a provider for a type.
+     * It will execute in the same way as an on<Trigger<T>> statement except that when it executes will be
+     * determined by the Director. This ensures that it will only run when it has permission to run and will
+     * transition between providers in a sensible way.
+     *
+     * @tparam T the provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Provides : public commands::ProviderBase<T, commands::ProviderAction::NORMAL> {};
+
+    /**
+     * The When DSL word is used to limit access to a provider unless a condition is true.
+     * This will prevent a provider from running if the state is false.
+     * However if there is a Entering<T> or Leaving reaction that has a Causing relationship for this When, it will
+     * be executed to ensure that the Provider can run.
+     *
+     * @tparam T the condition expression that must be true in order to execute this task.
+     */
+    template <typename State, template <typename> class expr, enum State::Value value>
+    struct When {
+
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+
+            // Tell the director about this when condition
+            reaction->reactor.powerplant.emit(std::make_unique<commands::WhenExpression>(
+                reaction,
+                typeid(State),
+                [](const int& v) { return expr<int>()(v, value); },
+                []() { return NUClear::dsl::operation::CacheGet<State>::get(); },
+                [](NUClear::Reactor& reactor,
```

Comment:
there are a few moving parts here. Could you please add a couple of inline comments

Reply:
Documented them on the constructor itself and here